### PR TITLE
Buildpacks CI/CD Implementation

### LINKED ISSUE #844: [DevOps] Utilize buildpacks in Circle CI deployments
**Description:**

Once we have completed implementation of scripts that can be used to deploy both the frontend and backend to Cloud.gov as buildpacks we need to update the Circle CI configuration to utilize the buildpack deployments instead of the pre-existing docker based deploy script.

**Acceptance Criteria:** 

- [x] Circle CI deploys to Cloud.gov using the new buildpack scripts
- [x] Deployment arguments are dynamically pulled in as environment variables from the Circle CI host.
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 
- [x] Update `.circleci/config.yml` to utilize the buildpack deployment scripts
- [x] Update `.circleci/config.yml` to pass in environment variables for the relevant target deploy environment.
- [x] Test deployment from Circle CI by using the label `Deploy with CircleCI`

**Supporting Documentation:**

**Open Questions:**

**Time Estimate:** 1-2 days

### PR DESCRIPTION
**Pull request by jtwillis92**
### Summary of Changes
Closes #844 

* Refactors the Circle CI config.yml to utilize re-usable commands for many common deployment steps
* Added Circle CI `job` configuration for deployment environments of dev, staging and prod
* Utilizes new buildpack deployment scripts instead of previous docker based deployment script
* Updates GitHub Action for label based deployment to call the new `deploy-dev` job
* Updates docker image used for running deployment commands to a non-legacy CircleCI image

### How to Test
* Delete the existing `tdp-backend` and `tdp-frontend` apps from the Cloud.gov Dashboard - Docker based app deployments cannot be converted to buildpack deployments as part of a rolling deploy by design of Cloud.gov. Delete only the apps and routes, leave all services as we do not yet provision those automatically.
* Assign the label `Deploy with CircleCI` to this PR
* Observe the deploy-dev job output on CircleCI and confirm that the deployment is successful
* Using the [Cloud Foundry CLI](https://docs.cloudfoundry.org/cf-cli/install-go-cli.html) extract the updated JWT_CERT:
```
cf login --sso
cf env tdp-backend | sed -n -e '/JWT_CERT/,/END CERTIFICATE/ p'
```
* Login to the [Login.gov Sandbox](https://dashboard.int.identitysandbox.gov/) and update the certificate there with the output from the step above. ([see also](https://github.com/raft-tech/TANF-app/blob/raft-tdp-main/docs/Technical-Documentation/jwt-key-rotation.md#devstaging-environments))
* Navigate to the deployed frontend application at `https://tdp-frontend.apps.cloud.gov` and confirm that the site works as intended.

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
  - [x] Circle CI deploys to Cloud.gov using the new buildpack scripts
  - [x] Deployment arguments are dynamically pulled in as environment variables from the Circle CI host.
  - [x] Testing Checklist has been run and all tests pass
  - [x] README is updated, if necessary
As facilitator/product manager, @kniz-raft will decide if ACs are met from Raft's perspective.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

N/A CI changes only

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Does this PR change any linting or CI settings?

### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

N/A No UX changes

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

+ [x] If this PR introduces backend code, is that code documented both inline and overall?
+ [x] If this PR introduces frontend code, is that code documented both inline and overall?
+ [x] If this PR introduces dependencies, are their licenses documented?

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,6 +1,63 @@
...
   node: circleci/node@4.2.1
+commands:
+  set-environment:
+    parameters:
+      env:
+        default: "DEV"
+        type: "string"
+    steps:
+      - run: |
+        CF_SPACE=CF_SPACE_<<parameters.env>>
```
Comment (jtwillis92) [inline .circleci/config.yml:12]: CF_SPACE wouldn't fit this pattern - that doesn't get defined in CircleCI but is solely based on the `$CIRCLE_BRANCH` 
ie:
HHS:main => tanf-prod
raft-tech:raft-tdp-main => tanf-dev (will be tanf-staging once we have #847 completed)
all others => tanf-dev

File: .circleci/config.yml [hunk 2]
```diff
@@ -1,6 +1,63 @@
...
+        CF_USERNAME=CF_USERNAME_<<parameters.env>>
+        CF_PASSWORD=CF_PASSWORD_<<parameters.env>>
+        CGHOSTNAME_FRONTEND=CGHOSTNAME_FRONTEND_<<parameters.env>>
+        CGHOSTNAME_BACKEND=CGHOSTNAME_BACKEND_<<parameters.env>>
+  deploy-frontend:
+    parameters:
+      deploy-strategy:
+        default: "rolling"
+        type: string
+      deploy-env:
```
Comment (jtwillis92) [inline .circleci/config.yml:22]: This variable never actually gets used in either of the deploy scripts - I think it would be safe to remove.

Setting the space with the `cf target` command should suffice for what I think this was intended for.

File: .circleci/config.yml [hunk 3]
```diff
@@ -1,6 +1,63 @@
...
+        default: "staging"
+        type: string
+      backend-hostname:
+        default: "https://tdp-backend.app.cloud.gov/"
```
Comment (jtwillis92) [inline .circleci/config.yml:26]:
```suggestion
        default: "tdp-backend"
```
looks like the way this gets used it should be just this portion of the hostname, perhaps CG_APPNAME is a better name for the variable.

File: .circleci/config.yml (hunk 1 above)
Comment (riatzukiza) [reply .circleci/config.yml]: I was planning to add new ENV vars to circle ci that matched this pattern so I wouldn't need to do condition checking to insert these values

File: .circleci/config.yml (hunk 2 above)
Comment (riatzukiza) [reply .circleci/config.yml:22]: 👍

File: scripts/deploy-frontend.sh [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+
+# perform a rolling update for the backend and frontend deployments if specifed,
+# otherwise perform a normal deployment 
+if [ $DEPLOY_STRATEGY = "rolling" ] ; then
+
+	  update_frontend 'rolling'
+else
+	  update_frontend
+fi
```
Comment (jorgegonzalez) [inline scripts/deploy-frontend.sh:59]: Uber nit but this file is missing tabs and spaces.

File: tdrs-frontend/src/components/SplashPage/SplashPage.jsx [hunk 5]
```diff
@@ -34,7 +34,7 @@ function SplashPage() {
           <div className="usa-hero__callout">
             <h1 className="usa-hero__heading">
               <span className="usa-hero__heading--alt font-serif-2xl margin-bottom-5">
-                Sign into TANF Data Portal
+                Sign into TANF Data Portal!
```
Comment (carltonsmith) [inline tdrs-frontend/src/components/SplashPage/SplashPage.jsx:37]: Curious why this change is happening here?

Comment (riatzukiza) [reply tdrs-frontend/src/components/SplashPage/SplashPage.jsx:37]: So I could look at the front page and see a change from circle ci deploying successfully

Comment (carltonsmith) [reply tdrs-frontend/src/components/SplashPage/SplashPage.jsx:37]: I see, that makes sense. I presume we will be removing this before putting it into final review?

**Comment by riatzukiza** [comment]
Need to add clam av to these commands before this can be accepted or we will have a regresion.

**Review by riatzukiza — APPROVED** [review]

File: scripts/deploy-frontend.sh [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+CGHOSTNAME_BACKEND=${3}
+
+update_frontend()
+{
+    echo DEPLOY_STRATEGY: "$DEPLOY_STRATEGY"
+    echo FRONTEND_HOST: "$CGHOSTNAME_FRONTEND"
+    echo BACKEND_HOST: "$CGHOSTNAME_BACKEND"
+    cd tdrs-frontend || exit
+
+    echo "REACT_APP_BACKEND_URL=https://$CGHOSTNAME_BACKEND.app.cloud.gov/v1" >> .env.production
```
Comment (ADPennington) [inline scripts/deploy-frontend.sh:19]: i'd like to better understand what's happening here re: `.env.production` @jtwillis92

File: scripts/set-backend-env-vars.sh [hunk 7]
```diff
@@ -0,0 +1,50 @@
+
+set -o allexport
+[[ -f ./tdrs-backend/.env.production ]] && source ./tdrs-backend/.env.production
```
Comment (ADPennington) [inline scripts/set-backend-env-vars.sh:3]: same q re: `.env.production`. want to be sure I understand where the keys are. I believe our goal is to leverage cloud.gov to get and set any secret keys, yes?

File: scripts/deploy-frontend.sh (hunk 6 above)
Comment (jtwillis92) [reply scripts/deploy-frontend.sh:19]: In the context of the frontend app the `.env.production` file is used to provide build time overrides for settings committed to the repo in the `.env` files for the frontend. This is distinct from the one used for the backend in that it does not contain secrets.

For React the environment variables are evaluated and injected to the code at build time so the environment variables set on the Cloud.gov application instance have no impact on a running site. We have to do this prior to running `yarn build` so that the correct URLs are in place but those URLs change between different deployment targets so we dynamically assign it based on the supplied arguments to the script.

Reference docs: https://create-react-app.dev/docs/adding-custom-environment-variables/#what-other-env-files-can-be-used

File: scripts/set-backend-env-vars.sh (hunk 7 above)
Comment (jtwillis92) [reply scripts/set-backend-env-vars.sh:3]: This is only used if you're running the script locally - in the CI context it won't have that file and will skip sourcing the environment variables, instead using what is set in Circle CI project settings.

File: scripts/deploy-frontend.sh (hunk 6 above)
Comment (jtwillis92) [reply scripts/deploy-frontend.sh:19]: As an additional note - the environment variables set on the machine performing the `yarn build` step always override anything stored in any `.env` file, as mentioned in the docs linked above:
> These variables will act as the defaults if the machine does not explicitly set them.

See also this [further reference on dotenv](https://github.com/motdotla/dotenv#what-happens-to-environment-variables-that-were-already-set) - which is what is used by React to process these variables.

In the raft-tech CircleCI environment variables we had `REACT_APP_BACKEND_HOST` and `REACT_APP_BACKEND_URL` set to the localhost URL used for testing - which was overriding the build stage environment variables placed into `.env.production`. Since we already have a `.env.test` file configured for CI and checked in to the repo we do not need to define those in Circle CI at all - so they have since been removed.

File: scripts/set-backend-env-vars.sh (hunk 7 above)
Comment (alexsoble) [reply scripts/set-backend-env-vars.sh:3]: @jtwillis92 @ADPennington I remember during Dev Sync yesterday we decided to change this line, but I don't remember how exactly we planned to change it ... possibly just remove it, since our preference would be not to store keys locally if possible?

Comment (jtwillis92) [reply scripts/set-backend-env-vars.sh:3]: Ah yeah - we had discussed removing this and adding documentation to the README on how to get environment variables from cloud.gov if needed.

Comment (jtwillis92) [reply scripts/set-backend-env-vars.sh]: Removed!

**Review by ADPennington — APPROVED** [review]
👍🏾 👍🏾 thanks @jtwillis92 and @riatzukiza! big milestone! 

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
  + [x] Circle CI deploys to Cloud.gov using the new buildpack scripts
  + [x]  Deployment arguments are dynamically pulled in as environment variables from the Circle CI host.
  + [x] Testing Checklist has been run and all tests pass
  + [x] README is updated, if necessary

As Product Owner, @lfrohlich will decide if ACs are met.

### Deliverable 2: Tested Code

_n/a: ci changes only_

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Does this PR change any linting or CI settings?

### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

_n/a_

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

_well-documented!_

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?
_no security issues detected_
